drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +111,251 @@ fn main() -> ! {
-                    // same time it is retrieving a short response from us.
-                    //
-                    // We keep FIFOWR full with zeros when we run out of Tx data
-                    // in order to avoid Tx underrun errors.
-                    //
-                    // So, there will always be "unsent" bytes in FIFOWR.
-                    //
-                    // Actual transmitted bytes are always going to equal the
-                    // number of received bytes. Since the rx_count is not
-                    // skewed by the Tx trailing bytes still in FIFOWR, just
-                    // use rx_count for both.
-
-                    // Update Tx state and account for trailing bytes
-                    // left in FIFOWR if present.
-                    //
-                    // If we cared about those remaining FIFOWR bytes:
-                    // let txremainder = fifostat.txlvl().bits() as usize;
-                    break 'outer;
+            }
+
+            // Let's check for any problems
+            let fifostat = self.spi.fifostat();
+
+            if fifostat.txerr().bit() {
+                // We don't do anything with tx errors other than record them
+                // The SP will see a checksum error if this is a reply, or the
+                // underrun happened after the number of reply bytes and it
+                // doesn't matter.
+                self.spi.txerr_clear();
+                self.stats.tx_underrun = self.stats.tx_underrun.wrapping_add(1);
+            }
+
+            if fifostat.rxerr().bit() {
+                // Rx errors are more important. They mean we're missing
+                // data. We should report this to the SP. This can be used to
+                // potentially throttle sends in the future.
+                self.spi.rxerr_clear();
+                self.stats.rx_overrun = self.stats.rx_overrun.wrapping_add(1);
+                err = Some(IoError::Flow);
+            }
+
+            // Are we done?
+            if csn_deasserted {
+                self.spi.ssd_clear();
+
+                if rx_msg.is_empty() {
+                    // This was a CSn pulse
+                    self.stats.csn_pulses =
+                        self.stats.csn_pulses.wrapping_add(1);
+                    err = Some(IoError::Flush);
                 }
-            } // FIFO polling loop
+
+                break;
+            }
+
+            // Read the CSn flag *after* checking if CSn is deasserted.
+            //
+            // This ordering allows us to loop one more time to exchange data
+            // and empty our fifos. Putting the check at the end also minimizes
+            // the time to access the fifos the first time through the loop.
+            // The goal here is to reduce overrun/underrun after already
+            // waiting for the CSn asserted interrupt to fire.
+            csn_deasserted = self.spi.ssd();
```

Comment:
I'm not sure that's important. I think either a destructive read would also work here since we will exit the loop the next time through before we read again.